perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_tcsMails.java

```diff
@@ -0,0 +1,161 @@
+
+	private final static Logger log = LoggerFactory.getLogger(urn_perun_user_attribute_def_virt_tcsMails.class);
+
+	@Override
+	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) throws InternalErrorException {
+
+		Attribute attribute = new Attribute(attributeDefinition);
+		Set<String> tcsMailsValue = new HashSet<>();
+
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_preferredMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_ISMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_publicAliasMails));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_privateAliasMails));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_o365EmailAddressesMU));
+
+		attribute.setValue(new ArrayList(tcsMailsValue));
+
+		return attribute;
+	}
+
+	@Override
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		// handle source user attributes changes
+		if (message instanceof AttributeSetForUser && isAffectedAttribute(((AttributeSetForUser) message).getAttribute().getFriendlyName())) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AttributeSetForUser) message).getUser()));
+
+		} else if (message instanceof AttributeRemovedForUser && isAffectedAttribute(((AttributeRemovedForUser) message).getAttribute().getFriendlyName())) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AttributeRemovedForUser) message).getUser()));
+
+		} else if (message instanceof AllAttributesRemovedForUser) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AllAttributesRemovedForUser) message).getUser()));
+
+		}
+
+		return resolvingMessages;
+	}
+
+	/**
+	 * Return true if attribute name is one of affected attributes in this virtual module.
+	 *
+	 * @param nameOfAttribute name of attribute to check
+	 * @return true if attribute is one of affected, false otherwise
+	 */
+	private boolean isAffectedAttribute(String nameOfAttribute) {
+		if(preferredMailFriendlyName.equals(nameOfAttribute)) return true;
+		else if(isMailFriendlyName.equals(nameOfAttribute)) return true;
```

Comment:
Since you end each branch with single ```return true;``` you can probably remove ``else`` from all branches and simply return ``false`` at the end if no previous ``if`` was true.

Reply:
Changed.